compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -783,6 +783,19 @@ trait Checking {
       else if (called.is(Trait) && !caller.mixins.contains(called))
         ctx.error(i"""$called is already implemented by super${caller.superClass},
                    |its constructor cannot be called again""", call.sourcePos)
+
+      if (caller.is(Module)) {
+        val objRef = caller.sourceModule.termRef
+        val traverser = new TreeTraverser {
+          def traverse(tree: Tree)(implicit ctx: Context) = tree match {
+            case tree: RefTree if tree.isTerm && tree.tpe <:< objRef =>
```

Comment:
You can avoid the subtype check here:
```suggestion
            case tree: RefTree if tree.isTerm && (tree.tpe.widen.classSymbol eq caller) =>
```

Reply:
Avoiding subtype check will ignore the prefix to the object, which might be problematic if the object is nested inside classes.

However, I couldn't find a valid program that exploits the prefix,  so maybe it's safe to do the optimization.